File: ontrack-kdsl/src/main/java/net/nemerosa/ontrack/kdsl/spec/extension/av/AutoVersioningAuditMgt.kt
package net.nemerosa.ontrack.kdsl.spec.extension.av

import com.apollographql.apollo.api.Optional
import net.nemerosa.ontrack.kdsl.connector.Connected
import net.nemerosa.ontrack.kdsl.connector.Connector
import net.nemerosa.ontrack.kdsl.connector.graphql.schema.AutoVersioningAuditEntriesQuery
import net.nemerosa.ontrack.kdsl.connector.graphql.schema.AutoVersioningAuditEntryByIdQuery
import net.nemerosa.ontrack.kdsl.connector.graphqlConnector

/**
 * Management interface for the audit of the auto versioning
 */
class AutoVersioningAuditMgt(connector: Connector) : Connected(connector) {

    /**
     * Gets an entry by ID
     */
    fun findEntryById(uuid: String): AutoVersioningAuditEntry? =
        graphqlConnector.query(
            AutoVersioningAuditEntryByIdQuery(uuid)
        )?.autoVersioningAuditEntries?.pageItems?.firstOrNull()
            ?.autoVersioningAuditEntryFragment
            ?.toAutoVersioningAuditEntry()

    /**
     * Gets the list of auto versioning audit entries
     */
    fun entries(
        offset: Int = 0,
        size: Int = 10,
        source: String? = null,
        project: String,
        branch: String? = null,
        version: String? = null,
    ): List<AutoVersioningAuditEntry> {
        return graphqlConnector.query(
            AutoVersioningAuditEntriesQuery(
                offset = Optional.present(offset),
                size = Optional.present(size),
                source = Optional.presentIfNotNull(source),
                project = Optional.present(project),
                branch = Optional.presentIfNotNull(branch),
                version = Optional.presentIfNotNull(version),
            )
        )?.autoVersioningAuditEntries?.pageItems?.map { item ->
            item.autoVersioningAuditEntryFragment.toAutoVersioningAuditEntry()
        } ?: emptyList()
    }

}

File: ontrack-web-tests/ontrack/extensions/auto-versioning/AutoVersioningAuditMgt.js
import {graphQLCall} from "@ontrack/graphql";
import {gql} from "graphql-request";
import {AutoVersioningAuditEntry} from "@ontrack/extensions/auto-versioning/AutoVersioningAuditEntry";
import {AutoVersioningAuditOrder} from "@ontrack/extensions/auto-versioning/AutoVersioningAuditOrder";
import {AutoVersioningAuditState} from "@ontrack/extensions/auto-versioning/AutoVersioningAuditState";

export class AutoVersioningAuditMgt {
    constructor(ontrack) {
        this.ontrack = ontrack
    }

    async entries({source, project, branch, version}) {
        const data = await graphQLCall(
            this.ontrack.connection,
            gql`
                query AutoVersioningAuditEntries(
                    $source: String,
                    $project: String,
                    $branch: String,
                    $version: String,
                ) {
                    autoVersioningAuditEntries(filter: {
                        source: $source,
                        project: $project,
                        branch: $branch,
                        version: $version,
                    }) {
                        pageItems {
                            order {
                                uuid
                            }
                            mostRecentState {
                                state
                            }
                        }
                    }
                }
            `,
            {
                source,
                project,
                branch,
                version,
            }
        )
        return data.autoVersioningAuditEntries.pageItems.map(entry => {
            return new AutoVersioningAuditEntry({
                order: new AutoVersioningAuditOrder(entry.order),
                mostRecentState: new AutoVersioningAuditState(entry.mostRecentState),
            })
        })
    }
}